From b3b79cada08dc8cd77d50d8b6e7dc3f54c719442 Mon Sep 17 00:00:00 2001
From: liwei <liwei1234@oppo.com>
Date: Thu, 20 Nov 2025 15:54:37 +0800
Subject: [PATCH 14/15] ANDROID: mm: skip oom killer to keep important task
 alive

Add hook to skip out of memory kill, ensure that allow certain important threads to keep alive.

Bug: 460636204
Bug: 472245484
Change-Id: I67433d13dfaaa301fa73b87530aaca7c37e2d74b
Signed-off-by: liwei <liwei1234@oppo.com>
(cherry picked from commit 9f81c2665ac3833307838ce0216bcab6f31a5933)
---
 drivers/android/vendor_hooks.c | 1 +
 include/trace/hooks/mm.h       | 3 +++
 mm/oom_kill.c                  | 5 +++++
 3 files changed, 9 insertions(+)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index 4d06fc8ce709..46762eb13772 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -578,6 +578,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_group_exit);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_init);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_add_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_finish);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_evaluate_task_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_swapmem_gather_init);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_oom_swapmem_gather_finish);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_rvh_set_gfp_zone_flags);
diff --git a/include/trace/hooks/mm.h b/include/trace/hooks/mm.h
index 09422699c98a..7c2e93f06035 100644
--- a/include/trace/hooks/mm.h
+++ b/include/trace/hooks/mm.h
@@ -529,6 +529,9 @@ DECLARE_HOOK(android_vh_swapmem_gather_add_bypass,
 DECLARE_HOOK(android_vh_swapmem_gather_finish,
 	TP_PROTO(struct mm_struct *mm),
 	TP_ARGS(mm));
+DECLARE_HOOK(android_vh_oom_evaluate_task_bypass,
+	TP_PROTO(struct task_struct *task, struct oom_control *oc, bool *bypass),
+	TP_ARGS(task, oc, bypass));
 DECLARE_HOOK(android_vh_oom_swapmem_gather_init,
 	TP_PROTO(struct mm_struct *mm),
 	TP_ARGS(mm));
diff --git a/mm/oom_kill.c b/mm/oom_kill.c
index 158a73c2e8b6..bfbffeb91346 100644
--- a/mm/oom_kill.c
+++ b/mm/oom_kill.c
@@ -318,6 +318,7 @@ static int oom_evaluate_task(struct task_struct *task, void *arg)
 {
 	struct oom_control *oc = arg;
 	long points;
+	bool bypass = false;
 
 	if (oom_unkillable_task(task))
 		goto next;
@@ -347,6 +348,10 @@ static int oom_evaluate_task(struct task_struct *task, void *arg)
 		goto select;
 	}
 
+	trace_android_vh_oom_evaluate_task_bypass(task, oc, &bypass);
+	if (bypass)
+		goto next;
+
 	points = oom_badness(task, oc->totalpages);
 	if (points == LONG_MIN || points < oc->chosen_points)
 		goto next;
-- 
2.50.1

